WordPress/gutenberg
# [RNMobile] Enable Third-Party Blocks to Set Aspect Ratio of Image via Inner Blocks

## Description

The React Native version of [Jetpack's Tiled Gallery block](https://jetpack.com/support/tiled-galleries/) will be using [the Inner Block API](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/inner-blocks) to display images in a gallery. As part of this project, we need to change the aspect ratio of the images that are displayed.

It is likely that other third-party blocks will face similar use cases in the future if they're pulling in the image block as an inner block. With that in mind, and to help achieve our end goal, this PR introduces some changes that will enable any third-party block to define the aspect ratio of an image when used as an inner block.

_Note, the goal behind this PR _isn't_ to introduce Jetpack-specific code, but rather to introduce generic changes that will hopefully be useful or re-usable for other third-party blocks. We have tried to be mindful against introducing anything that is specific only to Jetpack to the Gutenberg codebase._

## Types of changes

Outlined below are specifics around the two main changes that this PR includes. They are the introduction of a new context, `fixedHeight`, to prevent an overlap with the core Gallery block's set height and a tweak to the image block's `shapeStyle` prop.

### Address Conflict with Core Gallery Block's Style 

* The [core Gallery block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/gallery) introduced [two block contexts](https://github.com/WordPress/gutenberg/blob/bee52e68292357011a799f067ad47aa1c1d710e1/docs/reference-guides/block-api/block-context.md) in order to get around the challenge of editing/styling the image block as an inner block. Those are `allowResize` and `imageCrop`. 
* The `imageCrop` context [translates to `resizeMode='cover'`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L633), which is what we need to use in order to change the inner image's aspect ratios.
* The use of `imageCrop`, meant [the image block's `hasImageContext` variable](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L517) began to evaluate as `true` for the Tiled Gallery block. This then meant [the `fixedHeight` style](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L636) was being applied to it, which prevented the aspect ratio of images from being changed. 
* With the above in mind, this PR introduces a [new `fixedHeight` context](https://github.com/WordPress/gutenberg/pull/35089/files#diff-aa7f347bbb8d1aa1c43808397ef62ef1d0b789773b127768c01c913d1a956e19R96) to the Gallery block. As explained [here](https://github.com/WordPress/gutenberg/pull/35089/files#r717078745), this change prevents styles that are specific to the core Gallery block from impacting any block that uses the `imageCrop` context.

### Allow Third-Party Styles to be Passed

* The image block's [`shapeStyle` prop was updated](https://github.com/WordPress/gutenberg/pull/35089/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R698) to accept not only a string, but also full styles. This is to allow third-party blocks, the Tiled Gallery block in this case, to pass styles without needing to make edits directly to the image block's core files.

## How has this been tested?

As this PR is part of a wider effort to port the Jetpack Tiled Gallery block to Gutenberg Mobile, steps to specifically test the Jetpack block can be found in the Jetpack repository at https://github.com/Automattic/jetpack/pull/21166, with the central Gutenberg Mobile PR at https://github.com/wordpress-mobile/gutenberg-mobile/pull/4017.

For this specific PR, on the Gutenberg side, the following has been tested in the iOS and Android apps, with this branch checked out from view:

* Ensure the images in the Gallery block are a set height and that there are no regressions.
* Add an Image block and ensure the image displays at the correct size. Play around with the block's various settings, including the round style, and ensure there are no regressions. 

## Screenshots

These screenshots show our specific use cases for using contexts, which is to remove and style parts of the image block when used as an inner block. You'll see we were able to use contexts to make the images appear as squares:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/134989742-1b86fb15-1699-4b21-ae22-612527151a21.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/134991240-cfeac12b-4dd7-4466-b3ca-55948d09149a.png" width="100%"> |

## Checklist:
